File: python_cpp/uPIMulator_frontend/src/abi/word/_base_word.py
import math
from typing import List

from abi.word.representation import Representation
from encoder.byte import Byte


class BaseWord:
    def __init__(self, width: int):
        assert width > 0

        self._bits: List[bool] = [False for _ in range(width)]

    def width(self) -> int:
        return len(self._bits)

    def size(self) -> int:
        return self.width() // 8

    def sign_bit(self) -> bool:
        return self._bits[-1]

    def bit(self, i: int) -> bool:
        return self._bits[i]

    def set_bit(self, i: int) -> None:
        self._bits[i] = True

    def clear_bit(self, i: int) -> None:
        self._bits[i] = False

    def bit_slice(self, representation: Representation, begin: int, end: int) -> int:
        assert 0 <= begin < end <= self.width()

        slice_width = end - begin

        value = 0
        for i in range(slice_width):
            if self.bit(i + begin):
                if representation == Representation.SIGNED and i == slice_width - 1:
                    value -= 2 ** i
                else:
                    value += 2 ** i
        return value

    def set_bit_slice(self, begin: int, end: int, value: int) -> None:
        assert 0 <= begin < end <= self.width()

        if value >= 0:
            self._set_positive_bit_slice(begin, end, value)
        else:
            self._set_negative_bit_slice(begin, end, value)

    def value(self, representation: Representation) -> int:
        return self.bit_slice(representation, 0, self.width())

    def set_value(self, value: int) -> None:
        self.set_bit_slice(0, self.width(), value)

    def to_bytes(self) -> List[Byte]:
        num_bytes = math.ceil(self.width() / 8.0)

        bytes_: List[Byte] = []
        for i in range(num_bytes):
            begin = 8 * i
            end = min(8 * (i + 1), self.width())

            bytes_.append(Byte(self.bit_slice(Representation.UNSIGNED, begin, end)))
        return bytes_

    def from_bytes(self, bytes_: List[Byte]) -> None:
        for i, byte in enumerate(bytes_):
            begin = 8 * i
            end = min(8 * (i + 1), self.width())

            self.set_bit_slice(begin, end, byte.value())

    def _set_positive_bit_slice(self, begin: int, end: int, value: int) -> None:
        assert value >= 0

        slice_width = end - begin
        for i in range(slice_width):
            if value % 2:
                self.set_bit(i + begin)
            else:
                self.clear_bit(i + begin)

            value //= 2

        assert value == 0

    def _set_negative_bit_slice(self, begin: int, end: int, value: int) -> None:
        assert value < 0

        slice_width = end - begin

        self.set_bit(end - 1)
        value += 2 ** (slice_width - 1)

        if begin + 1 < end:
            self._set_positive_bit_slice(begin, end - 1, value)

File: python_cpp/uPIMulator_backend/src/abi/word/_base_word.cc
#include "abi/word/_base_word.h"

#include <algorithm>
#include <cassert>
#include <cmath>

namespace upmem_sim::abi::word {

_BaseWord::_BaseWord(int width) {
  assert(width > 0);
  bits_.resize(width);
}

int64_t _BaseWord::bit_slice(Representation representation, int begin,
                             int end) {
  assert(0 <= begin and begin < end and end <= width());
  assert(end - begin < 64);

  int slice_width = end - begin;
  int64_t value = 0;
  for (int i = 0; i < slice_width; i++) {
    if (bit(begin + i)) {
      if (representation == SIGNED and i == slice_width - 1) {
        value -= static_cast<int64_t>(pow(2, i));
      } else {
        value += static_cast<int64_t>(pow(2, i));
      }
    }
  }
  return value;
}

void _BaseWord::set_bit_slice(int begin, int end, int64_t value) {
  assert(0 <= begin and begin < end and end <= width());
  assert(end - begin < 64);

  if (value >= 0) {
    set_positive_bit_slice(begin, end, value);
  } else {
    set_negative_bit_slice(begin, end, value);
  }
}

encoder::ByteStream *_BaseWord::to_byte_stream() {
  int num_bytes = ceil(width() / 8.0);
  auto byte_stream = new encoder::ByteStream();
  for (int i = 0; i < num_bytes; i++) {
    int begin = 8 * i;
    int end = std::min(begin + 8, width());

    auto byte = static_cast<int>(bit_slice(UNSIGNED, begin, end));
    byte_stream->append(byte);
  }
  return byte_stream;
}

void _BaseWord::from_byte_stream(encoder::ByteStream *byte_stream) {
  for (int i = 0; i < byte_stream->size(); i++) {
    int begin = 8 * i;
    int end = std::min(begin + 8, width());

    int64_t byte = byte_stream->byte(i);
    set_bit_slice(begin, end, byte);
  }
}

void _BaseWord::set_positive_bit_slice(int begin, int end, int64_t value) {
  assert(0 <= begin and begin < end and end <= width());
  assert(value >= 0);

  int slice_width = end - begin;
  for (int i = 0; i < slice_width; i++) {
    if (value % 2) {
      set_bit(begin + i);
    } else {
      clear_bit(begin + i);
    }

    value /= 2;
  }

  assert(value == 0);
}

void _BaseWord::set_negative_bit_slice(int begin, int end, int64_t value) {
  assert(0 <= begin and begin < end and end <= width());
  assert(value < 0);

  set_bit(end - 1);

  if (begin + 1 < end) {
    int slice_width = end - begin;
    value += static_cast<int64_t>(pow(2, slice_width - 1));
    set_positive_bit_slice(begin, end - 1, value);
  }
}

}  // namespace upmem_sim::abi::word
